src/cinematic_waypoints/cinematic_controller.py

```diff
@@ -1,119 +1,134 @@
+    # If no bounding box is currently tracked, returns the largest bounding box in the list.
+    def identify_best_bb_match(self, bounding_boxes):
+        # If not already tracking a bounding box, return the largest box
+        if self.smoothed_bounding_box is None:
+            return CinematicController.identify_largest_bb(bounding_boxes)
+        centroid_importance = 1 - self.bb_match_dimension_importance
+
+        best_bb = None
+        min_error = math.inf
+        for bb in bounding_boxes:
+            # Use Euclidean distance of dimensions and centroid to compute the "error"
+            bb_dimensions = np.asarray(bb.get_dimensions())
+            dimension_error = numpy.linalg.norm(bb_dimensions - np.asarray(self.smoothed_bounding_box.get_dimensions()))
+            bb_centroid = np.asarray(bb.get_centroid())
+            centroid_error = numpy.linalg.norm(bb_centroid - np.asarray(self.smoothed_bounding_box.get_centroid()))
+
+            error = self.bb_match_dimension_importance * dimension_error + centroid_importance * centroid_error
+            if error < min_error:
+                min_error = error
+                best_bb = bb
+
+        return best_bb
+
+    # Returns the bounding box from the list provided with the largest area.
+    @staticmethod
+    def identify_largest_bb(bounding_boxes):
+        largest_bb = None
+        largest_area = -1
+        for bb in bounding_boxes:
+            bb_area = bb.get_area()
+            if bb_area > largest_area:
+                largest_area = bb_area
+                largest_bb = bb
+        return largest_bb
+
+    def update_latest_drone_state(self, drone_state):
+        self.latest_drone_state = drone_state
+
+    def update_cinematic_waypoints(self, cinematic_waypoints):
+        self.cinematic_waypoints = cinematic_waypoints
+
+    # Uses a geometric low-pass filter to smooth out human bounding boxes
+    def compute_low_pass_filter_bb(self):
+        latest_width, latest_height = self.latest_bounding_box.get_dimensions()
+        latest_x, latest_y = self.latest_bounding_box.get_centroid()
+
+        smoothed_width, smoothed_height = self.smoothed_bounding_box.get_dimensions()
+        smoothed_x, smoothed_y = self.smoothed_bounding_box.get_centroid()
+
+        new_width = smoothed_width * self.bb_filter_gamma + latest_width * (1 - self.bb_filter_gamma)
+        new_height = smoothed_height * self.bb_filter_gamma + latest_height * (1 - self.bb_filter_gamma)
+        new_x = smoothed_x * self.bb_filter_gamma + latest_x * (1 - self.bb_filter_gamma)
+        new_y = smoothed_y * self.bb_filter_gamma + latest_y * (1 - self.bb_filter_gamma)
+        return BoundingBox((new_width, new_height), (new_x, new_y))
+
+    def generate_waypoints(self):
+        if self.smoothed_bounding_box is None or self.latest_drone_state is None:
+            print("Warning: asked to generate waypoints, but some of the current states are None.")
+        if self.cinematic_waypoints==None:

```

Comment:
Minor point: in python, check if something is None by writing "if XXX is None" (not == None). See line 121 as an example.